game/layout/layout.py

```diff
@@ -0,0 +1,268 @@
+    unit_count: list[int] = field(default_factory=list)
+
+    # defintion which unit types are supported
+    unit_types: list[Type[DcsUnitType]] = field(default_factory=list)
+    unit_classes: list[UnitClass] = field(default_factory=list)
+    alternative_classes: list[UnitClass] = field(default_factory=list)
+
+    # Defines if this groupTemplate is required or not
+    optional: bool = False
+
+    # if enabled the specific group will be generated during generation
+    # Can only be set to False if Optional = True
+    enabled: bool = True
+
+    # TODO Caching for faction!
+    def possible_types_for_faction(self, faction: Faction) -> list[Type[DcsUnitType]]:
+        """TODO Description"""
+        unit_types = [t for t in self.unit_types if faction.has_access_to_dcs_type(t)]
+
+        alternative_types = []
+        for accessible_unit in faction.accessible_units:
+            if accessible_unit.unit_class in self.unit_classes:
+                unit_types.append(accessible_unit.dcs_unit_type)
+            if accessible_unit.unit_class in self.alternative_classes:
+                alternative_types.append(accessible_unit.dcs_unit_type)
+
+        if not unit_types and not alternative_types and not self.optional:
+            raise LayoutException
+
+        return unit_types or alternative_types
+
+    @property
+    def unit_counter(self) -> int:
+        """TODO Documentation"""
+        default = len(self.units)
+        if self.unit_count:
+            if len(self.unit_count) == 1:
+                count = self.unit_count[0]
+            else:
+                count = random.choice(range(min(self.unit_count), max(self.unit_count)))
+            if count > default:
+                logging.error(
+                    f"UnitCount for Group Layout {self.name} "
+                    f"exceeds max available units for this group"
+                )
+                return default
+            return count
+        return default
+
+    @property
+    def max_size(self) -> int:
+        return len(self.units)
+
+    def generate_units(
+        self, go: TheaterGroundObject, unit_type: Type[DcsUnitType], amount: int
+    ) -> list[TheaterUnit]:
+        """TODO Documentation"""
+        return [
+            TheaterUnit.from_template(i, unit_type, self.units[i], go)
+            for i in range(amount)
+        ]
+
+
+class TheaterLayout:
+    """TODO Documentation"""
+
+    def __init__(self, name: str, role: GroupRole, description: str = "") -> None:
+        self.name = name
+        self.role = role
+        self.description = description
+        self.tasks: list[GroupTask] = []  # The supported tasks
+        self.groups: list[GroupLayout] = []
+
+        # If the template is generic it will be used the generate the general
+        # UnitGroups during faction initialization. Generic Groups allow to be mixed
+        self.generic: bool = False
+
+    def usable_by_faction(self, faction: Faction) -> bool:
```

Comment:
I generally agree. But this check has to be done before the actual `ForceGroup` object gets created. Otherwise i would create a ForceGroup and check later if this group is usable or not which adds uneeded steps and logic to the system back again which we removed with the last review.
Dont get me wrong but in my mind this check here does not really add functionality to the layout it's just a simple "hey can you fit to this or not" and it just returns true or false. I agree on the general idea but this is the downside of checking beforehand if we can even use this one or not.

```python
for layout in db.LAYOUTS.layouts:
            if (
                layout.generic or layout in preset_layouts
            ) and layout.usable_by_faction(faction):
                # Creates a faction compatible GorceGroup
                self.add_or_update_force_group(ForceGroup.for_layout(layout, faction))
```

uses this and only creates a forcegroup if the layout can even be used by faction or not. For me this makes the handling aftwards thousand times easier as i dont have to check it again. I can just be sure "okay i can trust that this fits because we checked it earlier and we can progress further"